### PR TITLE
[fix] math Quaternion::fromMat3

### PR DESCRIPTION
**Pull request by lealzhan**
Re: #https://github.com/cocos/3d-tasks/issues/16899
Quaternion::fromMat3 copies implementation of **[glm::quat_cast](https://github.com/g-truc/glm/blob/master/glm/gtc/quaternion.inl)**

### Changelog

*

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/math/Quaternion.cpp [hunk 1]
```diff
@@ -370,46 +370,76 @@ void Quaternion::toEuler(const Quaternion &q, bool outerZ, Vec3 *out) {
     out->z = attitude;
 }
 
-void Quaternion::fromMat3(const Mat3 &m, Quaternion *out) {
-    CC_ASSERT(out);
+void Quaternion::fromMat3(const Mat3& m, Quaternion* out) {
```
Comment (minggo) [inline native/cocos/math/Quaternion.cpp]: Please don't change it. Just do it as other functions in this file.

File: native/cocos/math/Quaternion.cpp [hunk 2]
```diff
@@ -370,46 +370,76 @@ void Quaternion::toEuler(const Quaternion &q, bool outerZ, Vec3 *out) {
...
+	CC_ASSERT(out);
     float m00 = m.m[0];
-    float m10 = m.m[1];
-    float m20 = m.m[2];
-    float m01 = m.m[3];
+    float m01 = m.m[1];
+    float m02 = m.m[2];
+    float m10 = m.m[3];
```
Comment (minggo) [inline native/cocos/math/Quaternion.cpp:378]: It seems it is different of left hand or right hand usage.

Comment (lealzhan) [reply native/cocos/math/Quaternion.cpp:378]: the matrix is stored as colum major，here m01 means colum 0， row 1. the belowing is copied from glm

File: native/cocos/math/Quaternion.cpp (hunk 1 above)
Comment (lealzhan) [reply native/cocos/math/Quaternion.cpp]: ok

File: native/cocos/math/Quaternion.cpp [hunk 3]
```diff
@@ -371,45 +371,75 @@ void Quaternion::toEuler(const Quaternion &q, bool outerZ, Vec3 *out) {
...
+    float biggestVal = sqrt(fourBiggestSquaredMinus1 + 1) * 0.5F;
+	float mult = 0.25F / biggestVal;
+    switch(biggestIndex)
+    {
+    case 0:
+        out->w =  biggestVal;
+        out->x = (m12 - m21) * mult;
+        out->y = (m20 - m02) * mult;
+        out->z = (m01 - m10) * mult;
+        return;
```
Comment (minggo) [inline native/cocos/math/Quaternion.cpp]: It is better to use break.

File: native/cocos/math/Quaternion.cpp [hunk 4]
```diff
@@ -371,45 +371,75 @@ void Quaternion::toEuler(const Quaternion &q, bool outerZ, Vec3 *out) {
...
+        out->z = (m12 + m21) * mult;
+        return;
+    case 3:
+        out->w = (m01 - m10) * mult;
+        out->x = (m20 + m02) * mult;
+        out->y = (m12 + m21) * mult;
+        out->z = biggestVal;
+        return;
+    default: // Silence a -Wswitch-default warning in GCC. Should never actually get here. Assert is just for sanity.
+        assert(false);
```
Comment (minggo) [inline native/cocos/math/Quaternion.cpp]: Please use `CC_ASSERT`

File: native/cocos/math/Quaternion.cpp [hunk 5]
```diff
@@ -371,45 +371,75 @@ void Quaternion::toEuler(const Quaternion &q, bool outerZ, Vec3 *out) {
...
-        out->y = (m12 + m21) * s;
-        out->z = 0.25F / s;
+    float fourXSquaredMinus1 = m00 - m11 - m22;
+    float fourYSquaredMinus1 = m11 - m00 - m22;
+    float fourZSquaredMinus1 = m22 - m00 - m11;
+    float fourWSquaredMinus1 = m00 + m11 + m22;
+
+    int biggestIndex = 0;
+    float fourBiggestSquaredMinus1 = fourWSquaredMinus1;
+    if(fourXSquaredMinus1 > fourBiggestSquaredMinus1)
```
Comment (minggo) [inline native/cocos/math/Quaternion.cpp]: Should add a space between `if` and `{`.

File: native/cocos/math/Quaternion.cpp [hunk 6]
```diff
@@ -371,45 +371,75 @@ void Quaternion::toEuler(const Quaternion &q, bool outerZ, Vec3 *out) {
...
+    {
```
Comment (minggo) [inline native/cocos/math/Quaternion.cpp]: `{` should be the same line as `if`.

File: native/cocos/math/Quaternion.cpp [hunk 7]
```diff
@@ -371,45 +371,75 @@ void Quaternion::toEuler(const Quaternion &q, bool outerZ, Vec3 *out) {
 }
 
 void Quaternion::fromMat3(const Mat3 &m, Quaternion *out) {
```
Comment (stanleyljl) [inline native/cocos/math/Quaternion.cpp:373]: It is better to unity the ts & cpp implementation.

**Review by stanleyljl — APPROVED** [review]

**Review by yiwenxue — APPROVED** [review]

File: native/cocos/math/Mat3.cpp [hunk 8]
```diff
@@ -107,8 +107,8 @@ void Mat3::fromViewUp(const Vec3 &view, const Vec3 &up, Mat3 *out) {
     }
     Vec3::cross(view, vTempA, &vTempB);
     out->set(vTempA.x, vTempA.y, vTempA.z,
-             vTempB.x, vTempB.y, vTempB.z,
-             view.x, view.y, view.z);
+        vTempB.x, vTempB.y, vTempB.z,
+        view.x, view.y, view.z);
```
Comment (minggo) [inline native/cocos/math/Mat3.cpp]: I think previous format is better.

File: native/cocos/math/Quaternion.cpp (hunk 2 above)
Comment (minggo) [reply native/cocos/math/Quaternion.cpp:378]: Got it.

File: native/cocos/math/Quaternion.cpp (hunk 7 above)
Comment (lealzhan) [reply native/cocos/math/Quaternion.cpp:373]: ok

File: native/cocos/math/Quaternion.cpp (hunk 6 above)
Comment (lealzhan) [reply native/cocos/math/Quaternion.cpp]: ok

File: native/cocos/math/Quaternion.cpp (hunk 5 above)
Comment (lealzhan) [reply native/cocos/math/Quaternion.cpp]: ok

File: native/cocos/math/Quaternion.cpp (hunk 4 above)
Comment (lealzhan) [reply native/cocos/math/Quaternion.cpp]: ok

File: native/cocos/math/Quaternion.cpp (hunk 3 above)
Comment (lealzhan) [reply native/cocos/math/Quaternion.cpp]: ok

File: native/cocos/math/Mat3.cpp (hunk 8 above)
Comment (lealzhan) [reply native/cocos/math/Mat3.cpp]: ok

**Review by minggo — APPROVED** [review]

**Comment by lealzhan** [comment]
@cocos-robot run test cases

**Pull request merged**